test/unit/Cardano/Wallet/DB/MVarSpec.hs

```diff
@@ -71,44 +112,84 @@ dbReplaceValsProp db (key, val1, val2)  = monadicIO $ liftIO $ do
     putCheckpoint db key (initWallet val1)
     putCheckpoint db key (initWallet val2)
     resFromDb <- readCheckpoint db key
-
     resFromDb `shouldBe` (Just $ initWallet val2)
 
 dbMultiplePutsSeqProp
     :: DBLayer IO DummyState
-    -> KeyValPairs
+    -> KeyValPairs (PrimaryKey WalletId) DummyState
     -> Property
 dbMultiplePutsSeqProp db (KeyValPairs keyValPairs) = monadicIO $ liftIO $ do
     mapM_ (\(key, val) -> putCheckpoint db key (initWallet val)) keyValPairs
     resFromDb <- Set.fromList <$> readWallets db
-
     resFromDb `shouldBe` (Set.fromList (map fst keyValPairs))
 
 dbMultiplePutsParProp
     :: DBLayer IO DummyState
-    -> KeyValPairs
+    -> KeyValPairs (PrimaryKey WalletId) DummyState
     -> Property
 dbMultiplePutsParProp db (KeyValPairs keyValPairs) = monadicIO $ liftIO $ do
     mapConcurrently_
         (\(key, val) -> putCheckpoint db key (initWallet val))
         keyValPairs
     resFromDb <- Set.fromList <$> readWallets db
-
     resFromDb `shouldBe` (Set.fromList (map fst keyValPairs))
 
+dbMergeTxHistoryProp
+    :: DBLayer IO DummyState
+    -> KeyValPairs (PrimaryKey WalletId) (Map (Hash "Tx") (Tx, TxMeta))
+    -> Property
+dbMergeTxHistoryProp db (KeyValPairs keyValPairs) =
+    cover 90 cond "conflicting tx histories" prop
+  where
+    restrictTo k = filter ((== k) . fst)
+    -- Make sure that we have some conflicting insertion to actually test the
+    -- semantic of the DB Layer.
+    cond = map sumOfSizes ids /= map sizeOfUnion ids
+      where
+        ids = fst <$> keyValPairs
+        sizeOfUnion k = Map.size $ Map.unions $ snd <$> restrictTo k keyValPairs
+        sumOfSizes k = sum $ Map.size . snd <$> restrictTo k keyValPairs
+    prop = monadicIO $ liftIO $ do
+        forM_ keyValPairs $ \(key, val) -> do
+            putCheckpoint db key (initWallet $ DummyState 0)
```

Comment:
ok, so we are here putting tx histories using between (10,50) pairs, let's say X and having X wallet state checkpoints, each having `DummyState 0`. Then, for each key form X we read tx history. And expect that at each key the corresponding value is the same from res and when looking at keyValPairs (here Map.unions gather values if the key occurs more than once)